### PR TITLE
Add parallel extension methods to the `list` type

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Add parallel extension methods to the `list` type

Add some methods to `list` intended to make it easier to safely
modify lists in a parallel environment.

The `this()` accessor method is now disabled for lists that are
initialized with `parSafe=true`. The methods `getValue()` and
`getBorrowed()` have been added to make it possible to get a copy
of a list element when `parSafe=true`.

The `set()` method has been added to make it possible to modify the
value of a list element when `parSafe=true`.

The `update()` method has been way to access a reference to a list
element in a parallel-safe manner. It is intended for use with
lists that have been initialized with `parSafe=true`. While it can
be used with `parSafe=false` lists, use of `this()` should be
preferred as it is more idiomatic.

Tests have been added for each new method and the `this()` compiler
warning.

---

Testing:

- [x] ALL on linux64 when COMM=none
- [x] Docs rendered locally

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
Generated docs are here: http://web.us.cray.com/~dlongnecke/list-parsafe-extensions/index.html

@mppf @e-kayrakli @lydia-duncan @daviditen 

Would you please take a look at the following methods and let me know if their types and documentation seem agreeable to you?

- getValue
- set
- getBorrowed
- computeIndex

In particular computeIndex!

File: modules/standard/List.chpl
Comment (e-kayrakli) [inline modules/standard/List.chpl:1424]: does the error we get for not being able to resolve `worker.this` clean enough? Should we check whether we can resolve it and give a more helpful message?

File: modules/standard/List.chpl
Comment (e-kayrakli) [inline modules/standard/List.chpl:1445]: Same here: I am not sure whether this warning is necessary

Comment (e-kayrakli) [reply modules/standard/List.chpl:1445]: (which I see that this PR doesn't really add, but just change/adapt)

File: modules/standard/List.chpl
Comment (lydia-duncan) [inline modules/standard/List.chpl:1397]: I'm not sure that I understand what "in a guarded manner" means.  Based on just reading the docs, it seems like the amount of guarding would depend on the user function provided, but I think what you mean is that it will be performed in a parallel safe manner?  I'd probably say that instead

**Comment by mppf** [comment]
> The this() accessor method is now disabled when a list has been initialized with parSafe=true.

Also need to do this for `first()` and `last()`.

Re ` computeIndex` - I am happy with the functionality but I don't like the name. I'd just call it `update`, personally.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1330]: I think we have a precedent of warning about halts in these docs.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1424]: That's a great idea, I'll add a test with an unresolvable worker.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1397]: Thanks!

File: modules/standard/List.chpl
Comment (daviditen) [inline modules/standard/List.chpl:1377]: ... has been set ~~by~~ to `x`, ?

**Comment by dlongnecke-cray** [comment]
@mppf 

re: The name `computeIndex`

I would like to call it `update` as well, but the `map` type already has a method named `update` used to add another map's keys and values to an existing map. So for the sake of symmetry I thought I'd call both methods some variation of `compute` until the module review.

**Comment by mppf** [comment]
> I would like to call it update as well, but the map type already has a method named update used to add another map's keys and values to an existing map. So for the sake of symmetry I thought I'd call both methods some variation of compute until the module review.

I don't think that `update` is a reasonable term for that operation on the existing `map`. It should be renamed and then both `list` and `map` can have `update` for the new operation.

**Comment by dlongnecke-cray** [comment]
Do you think it's reasonable to rename that procedure for map when I add the "compute/update" method to it? Or should I wait until the module review for `map`?

**Comment by mppf** [comment]
I think we should rename the function while adding the compute/update method rather than creating one thing and then changing it in the module review. But you will have to check with a couple others to make sure I'm not the only one thinking in this way.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1377]: Thanks!

**Comment by dlongnecke-cray** [comment]
Hi @mppf, mind being the primary reviewer for this PR?

File: modules/standard/List.chpl
Comment (mppf) [inline modules/standard/List.chpl:1397]: You have to call `_destroy(dst)` before the move

File: modules/standard/List.chpl
Comment (mppf) [inline modules/standard/List.chpl:1421]: `worker` seems to me to be an odd name for this argument. `worker` might imply something about tasks to people.

File: modules/standard/List.chpl
Comment (mppf) [inline modules/standard/List.chpl:1446]: Should this and the error for first/last recommend using update or set or something? Additionally, since these used to work, do we need to deprecate it for now (i.e. make these a warning) rather than outright ban it?

**Review by mppf — COMMENTED** [review]
I'm happy with the general direction but there are some details to figure out before merging.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1392]: I was mimicking what we do for `map`, want to halt instead?

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1397]: Thanks!

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1421]: Suggestion? I'm drawing a blank.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1446]: Do deprecation warnings for `this`, `first`, and `last`, sound good?

File: modules/standard/List.chpl
Comment (mppf) [reply modules/standard/List.chpl:1421]: maybe "updater" or "operation" or "fn"

File: modules/standard/List.chpl
Comment (mppf) [reply modules/standard/List.chpl:1392]: I think for list it should behave similarly to how an out-of-bounds `this` worked, so yes, for now at least

Comment (mppf) [reply modules/standard/List.chpl:1392]: But I suppose this is one of those cases where there is a conflicting demand for consistency (between types and between fns).

File: modules/standard/List.chpl
Comment (mppf) [reply modules/standard/List.chpl:1446]: yes

Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1446]: I've gone with deprecation warnings and suggested using `set()` or `update()`.

File: modules/standard/List.chpl
Comment (dlongnecke-cray) [reply modules/standard/List.chpl:1421]: I've gone with the term `updater`, for lack of something better.

**Review by mppf — APPROVED** [review]

**Pull request merged**
